torch/csrc/DataLoader.cpp

```diff
@@ -0,0 +1,191 @@
+
+// In cases like data loader, if a worker process die due to bus error/segfault
+// or just hang, the main process, if implemented with
+// multiprocessing.queue.SimpleQueue, will hang waiting for data. This is
+// difficult to avoid on PyTorch side as it can be caused by limited shm, or
+// other libraries users call in the workers. The following methods is an effort
+// to do our best provide some error message to users when such unfortunate
+// events happen.
+
+// TODO: The following don't work on Windows. Specifically, waitpid calls and
+// SIGCHLD handler.
+
+#ifndef _WIN32
+
+// Critical signal handlers should be registered on worker processes before
+// doing work.
+// Python handle is _set_worker_signal_handlers().
+#define SIGNAL_HANDLER(SIGNAL, HANDLER_NAME, ERROR_MSG)                       \
+static void HANDLER_NAME(int sig)                                             \
+{                                                                             \
+    write(STDERR_FILENO, ERROR_MSG, sizeof(ERROR_MSG) / sizeof(char));        \
+    _exit(EXIT_FAILURE);                                                      \
+}
+
+// signal(2) is really not portable. So use sigaction.
+// http://man7.org/linux/man-pages/man2/signal.2.html
+#define SET_SIGNAL_HANDLER(SIGNAL, HANDLER_NAME, ERROR_FLAG)                  \
+{                                                                             \
+  struct sigaction sa;                                                        \
+  sa.sa_handler = HANDLER_NAME;                                               \
+  sa.sa_flags = SA_RESTART;                                                   \
+  sigemptyset(&sa.sa_mask);                                                   \
+  ERROR_FLAG |= sigaction(SIGNAL, &sa, NULL) != 0;                            \
+}
+
+SIGNAL_HANDLER(SIGBUS, handler_SIGBUS, "ERROR: Unexpected bus error encountered in worker. "
+  "This might be caused by insufficient shared memory (shm).\n");
+SIGNAL_HANDLER(SIGSEGV, handler_SIGSEGV, "ERROR: Unexpected segmentation fault encountered in worker.\n");
+
+static std::vector<pid_t> worker_pid_vec = {};
+// The following are needed since std::vector is not asynchronous safe.
+static std::atomic<pid_t *> worker_pids;
+static std::atomic<size_t> num_worker_pids(0);
+static std::atomic<size_t> num_dataloaders(0);
+
+PyObject *THPModule_setWorkerSignalHandlers(PyObject *module, PyObject *arg) {
+  HANDLE_TH_ERRORS
+  int error = 0;
+  SET_SIGNAL_HANDLER(SIGBUS, &handler_SIGBUS, error);
+  SET_SIGNAL_HANDLER(SIGSEGV, &handler_SIGSEGV, error);
+  if (error == 0) {
+    Py_RETURN_TRUE;
+  } else {
+    Py_RETURN_FALSE;
+  }
+  END_HANDLE_TH_ERRORS
+}
+
+// SIGCHLD hander should be registered on main loader process to catch any
+// worker failing. SIGALRM handler is needed for implementing timeout.
+// Python handles are _set_main_signal_handers() and
+// _remove_main_signal_handers().
+static void handler_SIGCHLD_main(int sig) {
```

Comment:
Does Python install it's own signal handlers? This doesn't look like it calls the previously installed signal handler.

Reply:
No python doesn't.

Reply:
There are a few hits for SIGCHLD in the Python source code that look relevant:

https://github.com/python/cpython/blob/fc6b348b12ad401cab0261b7b71a65c60a08c0a8/Lib/multiprocessing/forkserver.py#L173

https://github.com/python/cpython/blob/3e2ad8ec61a322370a6fbdfb2209cf74546f5e08/Lib/asyncio/unix_events.py#L793

https://github.com/python/cpython/blob/03eb11f0b354751248b427455b89e9340cfd2b47/Modules/posixmodule.c#L5926

Reply:
Some indeed look relevant :) Thanks, I'll make sure to call the original handler.